include/adm/detail/id_parser.hpp

```diff
@@ -83,5 +83,22 @@ namespace adm {
       const char *type;
       const std::string &id;
     };
+
+    /// write a hex value into an existing string
+    inline void formatHex(std::string &id, size_t start, size_t len,
+                          unsigned value) {
+      assert(start + len <= id.size());
+      for (int i = start + len - 1; i >= static_cast<int>(start); i--) {
```

Comment:
Does this provide any improvement over using a stringstream with std::hex? Might be more readable, but if it's slower (e.g, by allocating a stringstream object for a start) then not worth it.

Reply:
This is the only comment (and not a big deal in any case) - LGTM :+1:

Reply:
Thanks for going through this.

This is a good idea and I didn't test it. In general streams are quite slow (too much state manipulation) and it would definitely cause more allocation (the string can't be moved out of an ostringstream until c++20).

Here's how it looks -- format is the existing implementation, formatStream does this:

```cpp
namespace detail {
    inline void formatHex(std::ostream &s, size_t len, unsigned value) {
      auto pos_start = s.tellp();
      s << std::hex << std::setfill('0') << std::setw(len) << value;
      auto pos_end = s.tellp();

      if (pos_start + static_cast<decltype(pos_start)>(len) != pos_end) {
        std::ostringstream errorString;
        errorString << "failed to convert int value to hex: " << value;
        throw std::runtime_error(errorString.str());
      }
    }
}

std::string formatIdStream(const AudioBlockFormatId& id) {
  std::ostringstream s;
  s << "AB_";
  detail::formatHex(s, 4, id.get<TypeDescriptor>().get());
  detail::formatHex(s, 4, id.get<AudioBlockFormatIdValue>().get());
  s << "_";
  detail::formatHex(s, 8, id.get<AudioBlockFormatIdCounter>().get());

  return s.str();
}
```

```
benchmark name                       samples       iterations    estimated
                                     mean          low mean      high mean
                                     std dev       low std dev   high std dev
-------------------------------------------------------------------------------
format                                         100           574     1.3776 ms 
                                        22.4311 ns     22.371 ns    22.5304 ns 
                                       0.385954 ns   0.267418 ns    0.61799 ns 
                                                                               
formatStream                                   100            42     1.4028 ms 
                                        312.756 ns     310.67 ns    316.577 ns 
                                        13.9727 ns    9.25146 ns    25.3446 ns 
                                                                               
parse                                          100           724     1.3756 ms 
                                        18.3045 ns    18.2528 ns    18.3764 ns 
                                        0.30777 ns   0.230706 ns   0.512419 ns 
```

This would certainly not be a bottleneck when writing (it takes 1/10th of the time to serialise a document whereas the current version is less than 1/100th), but it's still relatively slower.

I'm fine with either really.